File: internal/app/repo/postgres/order_repo.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"mahir-trade-be/internal/app/repo/postgres/queries"

	"go.uber.org/dig"
)

type (
	OrderRepo interface {
		CreateOrder(ctx context.Context, req models.Order) (id int, err error)
		GetOrderByPaymentCode(ctx context.Context, paymentCode string) (resp models.Order, err error)
		UpdateOrderStatus(ctx context.Context, req models.Order) (err error)
	}

	OrderRepoImpl struct {
		dig.In

		*sql.DB
	}
)

func NewOrderRepo(impl OrderRepoImpl) OrderRepo {
	return &impl
}

func (o *OrderRepoImpl) CreateOrder(ctx context.Context, req models.Order) (id int, err error) {
	if req.Status == "" {
		req.Status = "pending"
	}

	rows, err := o.QueryContext(ctx, queries.QueryCreateOrder, req.UserID, req.PackageID, req.Status, req.PaymentCode, req.PaymentURL, req.CreatedBy)
	if err != nil {
		slog.ErrorContext(ctx, "[orderRepoImpl][CreateOrder] error while QueryContext", "%v", err.Error())
		return id, err
	}

	defer rows.Close()

	for rows.Next() {
		err = rows.Scan(&id)
		if err != nil {
			slog.ErrorContext(ctx, "[orderRepoImpl][CreateOrder] error while Scan", "%v", err.Error())
			return id, err
		}
	}

	return id, nil
}

func (o *OrderRepoImpl) GetOrderByPaymentCode(ctx context.Context, paymentCode string) (resp models.Order, err error) {
	row := o.QueryRowContext(ctx, queries.QueryGetOrderByPaymentCode, paymentCode)
	err = row.Scan(&resp.ID, &resp.UserID, &resp.PackageID, &resp.Status, &resp.PaymentCode, &resp.PaymentURL, &resp.CreatedBy, &resp.UpdatedBy, &resp.CreatedAt, &resp.UpdatedAt)
	if err != nil {
		if err == sql.ErrNoRows {
			return resp, fmt.Errorf("order with payment code %s not found", paymentCode)
		}

		slog.ErrorContext(ctx, "[orderRepoImpl][GetOrderByPaymentCode] error while Scan", "%v", err.Error())
		return resp, err
	}

	return resp, nil
}

func (o *OrderRepoImpl) UpdateOrderStatus(ctx context.Context, req models.Order) (err error) {
	_, err = o.ExecContext(ctx, queries.QueryUpdateOrderStatus, req.Status, req.UpdatedBy, req.ID)
	if err != nil {
		slog.ErrorContext(ctx, "[orderRepoImpl][UpdateOrderStatus] error while ExecContext", "%v", err.Error())
		return err
	}

	return nil
}

File: src/repo/order_repo.ts
import { Order } from "@models/order.model";
import db from "../db/db.config";
import { Pool } from "pg";
import { OrderQueries } from "./queries/order_queries";

export class OrderRepository {
  private pg: Pool;

  constructor() {
    this.pg = db;
  }

  // --- CREATE ORDER ---
  async createOrder(req: Order): Promise<number> {
    try {
      if (!req.status) {
        req.status = "pending";
      }

      const result = await this.pg.query<{ id: number }>(
        OrderQueries.CreateOrder,
        [
          req.user_id,
          req.package_id,
          req.status,
          req.payment_code,
          req.payment_url,
          req.created_by,
        ]
      );

      if (result.rowCount === 0) {
        throw new Error("failed to create order");
      }

      return result.rows[0].id;
    } catch (err: any) {
      console.error(`[orderRepo][createOrder] error: ${err.message}`);
      throw err;
    }
  }

  // --- GET ORDER BY PAYMENT CODE ---
  async getOrderByPaymentCode(paymentCode: string): Promise<Order> {
    try {
      const result = await this.pg.query<Order>(
        OrderQueries.GetOrderByPaymentCode,
        [paymentCode]
      );

      if (result.rowCount === 0) {
        throw new Error(`order with payment code ${paymentCode} not found`);
      }

      return result.rows[0];
    } catch (err: any) {
      console.error(`[orderRepo][getOrderByPaymentCode] error: ${err.message}`);
      throw err;
    }
  }

  // --- UPDATE ORDER STATUS ---
  async updateOrderStatus(req: Order): Promise<boolean> {
    try {
      const result = await this.pg.query(OrderQueries.UpdateOrderStatus, [
        req.status,
        req.updated_by,
        req.id,
      ]);

      return (result.rowCount ?? 0) > 0;
    } catch (err: any) {
      console.error(`[orderRepo][updateOrderStatus] error: ${err.message}`);
      throw err;
    }
  }
}
